### DISCUSSION TITLE
Allow u64 to be stored in an i64

### DISCUSSION THREAD

**Discussion by christian-oudard** [diesel-rs/diesel#3080] [Ideas]
Hello,

I am an engineer for MobileCoin. We are using SQLite, diesel, and rocket to build our wallet service.
https://github.com/mobilecoinofficial/full-service

In our database, we are storing amounts of MOB as u64 integers representing pico-MOB. For large values, we use SQLite negative values, which then get converted back after querying.

It would be very convenient to put some sort of filter at the database layer to convert i64 to u64 as they come out of the database. Unfortunately, I was unable to find any way that Diesel can do this conversion automatically, so we have been doing a lot of `as i64` and `as u64`.

The reason given in the compiler error why we can't have our model structs use u64 is because "the trait `diesel::Expression` is not implemented for `u64`". Is there a way to implement this, so that it mimics our strategy of converting u64 to i64 during storage? Is this something that would be useful in the base library, or would it just be application-specific?

**Comment by weiznich** [comment]
Diesel won't allow you using a `u64` in place of an `i64` for the same reasons that's not possible in rust. They are just not the same types. It's not up to discussion to change this mapping.

That written there are multiple ways you can change that in your application:
* You can use a new type wrapper around `u64` to implement a custom type mapping based on this.
* You can create a custom sql type, to mark columns that should allow `u64` values. Using this sql type you can define mappings from and to `u64` directly. (This will only work for non nullable use cases).
* You can use [`#[diesel(deserialize_as = "NewTypeWrapper")]`](https://docs.diesel.rs/1.4.x/diesel/deserialize/trait.Queryable.html) and [`#[diesel(serialize_as = "NewTypeWrapper")]`](https://docs.diesel.rs/master/diesel/prelude/derive.Insertable.html#optional-field-attributes) to make the first variant even better to use. (Note that the second attribute is only available on master yet.)
